src/Enzo/enzo_EnzoMethodPmDeposit.cpp

```diff
@@ -117,45 +118,41 @@ void EnzoMethodPmDeposit::compute ( Block * block) throw()
     block->lower(&xm,&ym,&zm);
     block->upper(&xp,&yp,&zp);
     block->cell_width(&hx,&hy,&hz);
-
-    // declare particle position arrays
-
-    ParticleDescr * particle_descr = cello::particle_descr();
-    Grouping * particle_groups = particle_descr->groups();
-
-    int num_mass = particle_groups->size("has_mass");
-
-    // Accumulate particle density using CIC
-
+    
+    // Get cosmological scale factors, if cosmology is turned on
     enzo_float cosmo_a=1.0;
     enzo_float cosmo_dadt=0.0;
     EnzoPhysicsCosmology * cosmology = enzo::cosmology();
-
+    
     if (cosmology) {
-
-      double time = block->time();
-      double dt   = block->dt();
-      cosmology->compute_expansion_factor (&cosmo_a,&cosmo_dadt,time+alpha_*dt);
-
+      cosmology->compute_expansion_factor(&cosmo_a,&cosmo_dadt,
+					  block->time() + alpha_*block->dt());
     }
-    if (rank >= 1) hx *= cosmo_a;
-    if (rank >= 2) hy *= cosmo_a;
-    if (rank >= 3) hz *= cosmo_a;
```

Comment:
It seems like that removing this is going to affect the way that the gas density gets deposited (which happens at the end of this function).

If this is true, then it's a little disturbing that none of our automated tests seem to detect this.

Reply:
Okay, I'll look into this. I was under the impression that that both gas density and particle density were comoving quantities. I suppose if a physical length scale is required for depositing the gas density, then I can define two separate variables i.e hx_com and hy_phys, or something similar.

Reply:
I was just trying to point out that this change was going to modify the deposititon of the gas density.

You could definitely be correct (I know basically nothing about the units of cosmological simulations). In that case the old implementation must have been wrong.

Could somebody familiar with the density units used for cosmological simulations in Enzo quickly clarify this for us? @jwise77 @bwoshea  @gregbryan

Reply:
Yes, I know what you mean. It would be good to understand which quantities need to be passed to dep_grid_cic (i.e., physical or comoving). I am also concerned that the value of dtf which is passed is not correct. Currently it is set to alpha_, which is a dimensionless parameter which is 0.5 by default, rather than a quantity with dimensions of time.

Reply:
In Enzo, the particle "mass" is actually a comoving density, adjusted for the level that it is on, so that when it gets deposited to the GravitatingMassField (which is the comoving density field used to calculate the gravitational acceleration), there is no multiplication by a volume. Here, I'm not quite sure what should be done -- the mass of a particle is a constant, so if you wanted to compute a comoving density, then you would want to divide by the comoving volume of the cell.  Normally dx is comoving, so a factor of the scale factor is not required, but I have to admit I have not really looked at what is going on in this routine, so best to take what I say with a grain of salt.

Reply:
@gregbryan , the purpose of this PR is to redefine the "mass" attribute, so that it is always treated as mass, rather than (sometimes) as density. As far as I know, in a cosmological simulation, length scales and densities are all comoving quantities, which is why I deleted the lines which multiply hx,hy,hz by "cosmo_a", so that, when the mass is multiplied by 1/(hx * hy * hz), the output is a comoving density.

The issue which @mabruzzo pointed out, is that this will affect the deposition of gas density, which is done by calling the 'dep_grid_cic' function:

`
FORTRAN_NAME(dep_grid_cic)(de,de_gas,temp,
  
  
      			       vx, vy, vz,
  
  
      			       &amp;dtf, rfield, &amp;rank,
  
  
      			       &amp;hxf,&amp;hyf,&amp;hzf,
  
  
      			       &amp;mx,&amp;my,&amp;mz,
  
  
      			       &amp;gxi,&amp;gyi,&amp;gzi,
  
  
      			       &amp;nxi,&amp;nyi,&amp;nzi,
  
  
      			       &amp;i0,&amp;i0,&amp;i0,
  
  
      			       &amp;nx,&amp;ny,&amp;nz,
  
  
      			       &amp;i1,&amp;i1,&amp;i1);`
      			       
Should hxf, hyf, hzf be comoving or physical length scales? 

I'm also surprised that dtf is set equal to 'alpha_'. Should it not be the same as the 'dt' used to drift the particles, i.e.:
`const double dt = alpha_ * block->dt() / cosmo_a;`

Reply:
I took a look at ``dep_grid_cic``, and I think that it's effectively just advancing the density over `dt` (i.e. the analogous operation to drifting the particles). I'm also pretty sure that the output density `de_gas` has the same units as the input `de`
- As an aside, I'm pretty sure that a lot of these arguments that get passed have to do with complications that arise from the patch-based AMR used in the original _Enzo_.

If I'm remembering correctly, the fields used to store gas properties store values in comoving units for cosmological simulations, so `hxf`, `hyf`, and `hzf` should probably be comoving.

Furthermore, since the `"density_total"` field is the sum of the `"density"` field (after drifting) and the `"density_particle"` field, I suspect that `"density_total"`, `"density_particle"` and `"density_particle_accumulate"` should all be comoving densities (in cosmological simulations).

I could be wrong, but this makes the most sense to me. The easiest way to check this would be to write some tests.

Reply:
Without doing a deep-dive back into this, it's unclear to me if you've made any changes on this front. If you have made changes, I'll take another look

Reply:
> I'm also surprised that dtf is set equal to 'alpha_'. Should it not be the same as the 'dt' used to drift the particles, i.e.:
const double dt = alpha_ * block->dt() / cosmo_a;

I agree that it should be the same dt as the particle drift, but in Enzo, dt is [actually zero](https://github.com/enzo-project/enzo-dev/blob/main/src/enzo/Grid_DepositBaryons.C#L198) for the PPM solver but not the RK2 solver ([diff from 2011](https://github.com/enzo-project/enzo-dev/commit/c16a9eb743d4716f487ac0ace62e4367d65b54aa#diff-0264e55b281d53b17e54c49abf1658d7ce6a19db6fe03686b7d225f1e9049908L174)).

I looked at a very old version of Enzo from 2005, and I found this comment that using a non-zero dt creates an instability. I think we should change it to zero to be consistent with Enzo.

```
    float dt = (DepositTime - Time)/a;
    dt = 0; /* the previous line is correct but creates an instability */
```

Reply:
OK, I looked this over and I *think* removing the multiplication by cosmo_a is the right thing to do. As others have pointed out, hx,hy,hz should be comoving, so this is not necessary.  Regarding alpha_, this is for the particles, and I think that is only set to zero for the baryons, so OK to keep as is?